resolwe_bio/processes/reads_processing/bbduk.yml

```diff
@@ -11,221 +11,363 @@
     expression-engine: jinja
     executor:
       docker:
-        image: resolwebio/legacy:1.0.0
-  data_name: '{{ reads.fastq.0.file|default("?") }}'
-  version: 0.0.6
+        image: resolwebio/rnaseq:3.0.0
+    resources:
+      cores: 10
+  data_name: '{{ reads|sample_name|default("?") }}'
+  version: 1.0.0
   type: data:reads:fastq:single:bbduk
   category: analyses
   flow_collection: sample
   persistence: CACHED
   description: >
-    BBduk is capable of quality-trimming and filtering, adapter-trimming, contaminant-filtering via
-    kmer matching, sequence masking, GC-filtering, length filtering, entropy-filtering, format
-    conversion, histogram generation, subsampling, quality-score recalibration, kmer cardinality
-    estimation, and various other operations in a single pass.
+    BBDuk combines the most common data-quality-related trimming, filtering, and masking operations
+    into a single high-performance tool. It is capable of quality-trimming and filtering,
+    adapter-trimming, contaminant-filtering via kmer matching, sequence masking, GC-filtering,
+    length filtering, entropy-filtering, format conversion, histogram generation, subsampling,
+    quality-score recalibration, kmer cardinality estimation, and various other operations in a
+    single pass. See
+    [the official website](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbduk-guide/)
+    for more information.
   input:
     - name: reads
       label: Reads
       type: data:reads:fastq:single
-    - name: adapters
-      label: Adapters
+    - name: show_advanced
+      label: Show advanced parameters
+      type: basic:boolean
+      default: false
+    - name: reference
+      label: Reference
+      hidden: '!show_advanced'
```

Comment:
(1) I ran BBDuk without any parameters to obtain the full set of supported parameters. Note that the version of BBDuk installed in our RNA-seq Docker image is newer than the version on GitHub that you linked which may be the reason why you couldn't find some parameters. Process inputs follow the same order as parameters displayed in the command line and are grouped similarly as well. I agree that matching command-line options with descriptions of inputs in the process is not always trivial. I can (a) add names of command-line parameters to descriptions or (b) change the names of process inputs to match the names of command-line parameters. However, (a) I didn't want the process input schema to look too technical and arouse a command-line-like feeling. Many of our users want to move away from the command line. (b) BBDuk uses a non-standard parameter naming conventions which doesn't play well with the rest of our processes. Lastly, imagine a refactoring where process inputs are recipes-like user-friendly and cannot be mapped one-to-one onto command-line parameters of the tool. I think that we should strive towards improving the descriptions of inputs to the point where the user would be able to use the process without the manual.

(2) This refactoring closely follows the function of the underlying tool. For example, if you call the process without parameters (except reads) it behaves similarly to the tool called from the command-line with just reads. This provides full control of operation and guarantees basic predictability of results: if an input in the process is not set explicitly, the tool's default value is used. This (a) eases integration into workflows because only the parameters explicitly set there have an effect on the operation of the tool and the process doesn't do anything unexpected automatically. This also (b) keeps power users in full control to achieve exactly what they want. The process is not intended to be used by basic users. For them we have recipes, workflows, and easy to use meta-processes named _for biologists_.

Custom input with adapters provided by BBDuk has been removed due to redundancy. The idea is to make those adapters available as public resources on our servers where they can be chosen as a regular data object by all trimming and filtering processes, including Trimmomatic and Cutadapt.

(3) If only reads are provided, they are copied to output and analyzed by FastQC. A warning is displayed that reads remain unmodified. This solution is based on what we've recently implemented in Trimmomatic and follows the operation of the tool itself which also prints out an equivalent warning.

Which parameters are important and which not is to some extent subjective so I decided to hide all non-essential parameters by default which gives a clean and non-intimidating interface to start with. To enable/show reference sequence-related parameters only after such sequences are provided is certainly an interesting idea. I base my decision not to implement this now on the fact that the process is much more likely to be used within a workflow than directly. Given the limited amount of time I deem other tasks more essential but I'm open to implementing this in the future.

Reply:
With a purpose of reproducibility of the BBduk results I think we should enable user to understand from the label or description which BBduk parameter is in use. Since there is so many parameters it is rather hard to find the one that user wish to define, if he has pre-defined command line pipeline. However I agree that many of our users want to move away from the command line. Why not adding at the end of description a sentence:"BBduk command line parameter `k`"? This is just my concern and suggestion based on user story, which I have described. The decision is up to you. 

To use tool from BioDash it would be much easier if not all parameters would be in advance part and there would be some parameter enabling. However I understand and agree this would take some additional time, and it is not a priority. I have added some of suggestions, where I found it most confusing...

About providing only reads file, I have the missed Warning message at first. I only added a question regarding `Minimum length` parameter.

Reply:
I added command-line parameter names in square brackets to avoid confusion with existing round ones in the labels. I followed your other suggestions as well.